src/users.py

```diff
@@ -0,0 +1,440 @@
+from collections import defaultdict
+from weakref import WeakSet
+import fnmatch
+import re
+
+from src.context import IRCContext, Features, lower
+from src.logger import debuglog
+from src import settings as var
+from src import db
+
+import botconfig
+
+Bot = None # bot instance
+
+all_users = WeakSet()
+
+_arg_msg = "(nick={0}, ident={1}, host={2}, realname={3}, account={4}, allow_bot={5})"
+
+predicate = re.compile(r"^[0-9]+$").search
+
+def get(nick=None, ident=None, host=None, realname=None, account=None, *, allow_multiple=False, allow_none=False, allow_bot=False, raw_nick=False):
+    """Return the matching user(s) from the user list.
+
+    This takes up to 5 positional arguments (nick, ident, host, realname,
+    account) and may take up to four keyword-only arguments:
+
+    - allow_multiple (defaulting to False) allows multiple matches,
+      and returns a list, even if there's only one match;
+
+    - allow_none (defaulting to False) allows no match at all, and
+      returns None instead of raising an error; an empty list will be
+      returned if this is used with allow_multiple;
+
+    - allow_bot (defaulting to False) allows the bot to be matched and
+      returned;
+
+    - raw_nick (defaulting to False) means that the nick has not been
+      yet parsed, and so ident and host will be None, and nick will be
+      a raw nick of the form nick!ident@host.
```

Comment:
It makes for a much cleaner and safer API to just enforce blah(hostmask) or blah(nick, ident, host) (no default value for ident/host so that you must pass all 3 if calling that overload). Since python doesn't have function overloading of that form (at least not in 3.4), they need to be separate function names instead.